File: call-api.py
#!/usr/bin/env python3
"""

call-api.py : Shadowserver Foundation API Utility

This script requires your API details to be stored in ~/.shadowserver.api 
with the following contents:

--
[api]
key = 123456798
secret = MySecret
uri = https://transform.shadowserver.org/api2/
--

This script may be called with two or three arguments:

    call-api.py <method> <request> [pretty|binary]

The request must be a valid JSON object.

Simple usage:

$ ./call-api.py test/ping '{}'
{"pong":"2020-10-26 23:06:37"}

Pretty output:

$ ./call-api.py test/ping '{}' pretty
{
    "pong": "2020-10-26 23:06:42"
}

"""

import os
import sys
import hmac
import hashlib
import json
import configparser

from urllib.request import urlopen, Request

config = configparser.ConfigParser()
config.read(os.environ['HOME'] + "/.shadowserver.api")

TIMEOUT = 45


def api_call(method, request):
    """
    Call the specified api method with a request dictionary.

    """

    url = config.get('api', 'uri') + method

    request['apikey'] = config.get('api', 'key')
    request_string = json.dumps(request)

    secret_bytes = bytes(str(config.get('api', 'secret')), 'utf-8')
    request_bytes = bytes(request_string, 'utf-8')

    hmac_generator = hmac.new(secret_bytes, request_bytes, hashlib.sha256)
    hmac2 = hmac_generator.hexdigest()

    ua_request = Request(url, data=request_bytes, headers={'HMAC2': hmac2})
    response = urlopen(ua_request, timeout=TIMEOUT)

    return response.read()


if __name__ == '__main__':

    if (len(sys.argv) < 3):
        exit("Usage: call-api.py method json [pretty|binary]")

    try:
        api_request = json.loads(sys.argv[2])
    except Exception as e:
        exit("JSON Exception: " + format(e))

    try:
        config.get('api', 'key')
    except configparser.NoSectionError:
        exit("Exception: " + os.environ['HOME'] + "/.shadowserver.api could "
             "not be found. Exiting.")

    try:
    	res = api_call(sys.argv[1], api_request)
    except Exception as e:
        exit("API Exception: " + format(e))

    if (len(sys.argv) > 3):
       if (sys.argv[3] == "pretty"):
           try:
               print(json.dumps(json.loads(res), indent=4))
           except:
               print(res.decode('utf-8'))
       elif (sys.argv[3] == "binary"):
           os.write(1, res);
       else:
           exit("Unknown option " + sys.argv[3])
    else:
        print(res.decode('utf-8'))

File: call-api.pl
#!/usr/bin/env perl

=head1 NAME

call-api.pl : Shadowserver Foundation API Utility

=head1 DESCRIPTION

This script requires your API details to be stored in ~/.shadowserver.api 
with the following contents:

--
[api]
key = 123456798
secret = MySecret
uri = https://transform.shadowserver.org/api2/
--

This script may be called with two or three arguments:

    call-api.pl <method> <request> [pretty|binary]

The request must be a valid JSON object.

Simple usage:

$ ./call-api.pl test/ping '{}' 
{"pong":"2020-10-26 23:06:37"}

Pretty output:

$ ./call-api.pl test/ping '{}' pretty
{
    "pong": "2020-10-26 23:06:42"
}

=cut

use Config::Simple;
use JSON;
use Digest::SHA qw(hmac_sha256_hex);
use LWP::UserAgent;
use HTTP::Request;
use URI;
use strict;

my $config = eval { new Config::Simple($ENV{'HOME'} . "/.shadowserver.api") };

my $TIMEOUT = 45;


=item api_call( $method, \%request )

Call the specified api method with a request dictionary.

=cut
sub api_call
{
	my ($method, $request) = @_;

	my $url = $config->param("api.uri") . $method;

	$request->{'apikey'} = $config->param('api.key');
    my $request_string = encode_json($request);
	my $hmac2 = hmac_sha256_hex($request_string, $config->param('api.secret'));

	my $ua = new LWP::UserAgent();
	$ua->timeout($TIMEOUT);

	my $ua_request = new HTTP::Request('POST', $url, [ 'HMAC2' => $hmac2 ] );
	$ua_request->content($request_string);
	my $response = $ua->request($ua_request);
	return $response->content;
}

unless (caller) # main
{

    if (scalar @ARGV < 2)
	{
        print STDERR "Usage: call-api.pl method json [pretty|binary]\n";
		exit(1);
	}

	my $api_request = eval { decode_json($ARGV[1]) };
	if ($@)
	{
		print STDERR "JSON Exception: $@\n";
		exit(1);
	}

	if (!defined($config) || $config->param('api.key') eq '')
	{
		print STDERR "Exception: api.key not defined in " . $ENV{'HOME'} 
			. "/.shadowserver.api\n";
		exit(1);
	}

    my $res = eval { api_call($ARGV[0], $api_request) };
	if ($@)
	{
		print STDERR "API Exception: $@\n";
		exit(1);
	}

    if (scalar @ARGV > 2)
	{
		if ($ARGV[2] eq "pretty")
		{
			eval { print to_json(decode_json($res), { pretty => 1}), "\n" };
			exit(0) unless ($@);
		}
    	elsif ($ARGV[2] eq "binary")
		{
			binmode(STDOUT);
			write(STDOUT, $res);
			exit(0);
		}
		else
		{
			die('Unknown option ' . $ARGV[2]);
		}
	}

	print $res, "\n";
}
